File: KEM/bindings/android/kazkem/src/main/kotlin/com/antrapol/kaz/kem/KazKemException.kt
package com.antrapol.kaz.kem

/**
 * Exception thrown by KAZ-KEM operations.
 *
 * @property errorCode Native error code
 * @property message Human-readable error message
 */
class KazKemException(
    val errorCode: Int,
    override val message: String
) : Exception(message) {

    companion object {
        // Error codes
        const val ERROR_SUCCESS = 0
        const val ERROR_INVALID_PARAM = -1
        const val ERROR_MEMORY = -2
        const val ERROR_RNG = -3
        const val ERROR_OPENSSL = -4
        const val ERROR_MSG_TOO_LARGE = -5
        const val ERROR_NOT_INIT = -6
        const val ERROR_INVALID_LEVEL = -7

        /**
         * Create exception from error code with descriptive message.
         */
        @JvmStatic
        fun fromErrorCode(errorCode: Int, operation: String = "Operation"): KazKemException {
            val description = when (errorCode) {
                ERROR_INVALID_PARAM -> "Invalid parameter"
                ERROR_MEMORY -> "Memory allocation failed"
                ERROR_RNG -> "Random number generation failed"
                ERROR_OPENSSL -> "OpenSSL error"
                ERROR_MSG_TOO_LARGE -> "Message too large"
                ERROR_NOT_INIT -> "KAZ-KEM not initialized"
                ERROR_INVALID_LEVEL -> "Invalid security level"
                else -> "Unknown error"
            }
            return KazKemException(errorCode, "$operation failed: $description (code: $errorCode)")
        }
    }

    /**
     * Check if this is a specific error type.
     */
    fun isNotInitialized(): Boolean = errorCode == ERROR_NOT_INIT
    fun isInvalidParameter(): Boolean = errorCode == ERROR_INVALID_PARAM
    fun isMemoryError(): Boolean = errorCode == ERROR_MEMORY
    fun isRngError(): Boolean = errorCode == ERROR_RNG

    override fun toString(): String = "KazKemException(code=$errorCode, message=$message)"
}

/**
 * Exception thrown when KAZ-KEM is not initialized.
 */
class NotInitializedException(message: String = "KAZ-KEM is not initialized. Call KazKem.initialize() first.") :
    Exception(message)

/**
 * Exception thrown for invalid parameter errors.
 */
class InvalidParameterException(message: String) : Exception(message)

File: KEM/bindings/dotnet/KazKem/src/KazKemException.cs
namespace Antrapol.Kaz.Kem;

/// <summary>
/// Exception thrown when a KAZ-KEM operation fails.
/// </summary>
public class KazKemException : Exception
{
    /// <summary>
    /// Native error code from the KAZ-KEM library.
    /// </summary>
    public int ErrorCode { get; }

    public KazKemException(string message) : base(message)
    {
        ErrorCode = -1;
    }

    public KazKemException(string message, int errorCode) : base(message)
    {
        ErrorCode = errorCode;
    }

    public KazKemException(string message, Exception innerException) : base(message, innerException)
    {
        ErrorCode = -1;
    }

    /// <summary>
    /// Create exception from native error code.
    /// </summary>
    internal static KazKemException FromErrorCode(int errorCode, string operation)
    {
        var message = errorCode switch
        {
            -1 => $"{operation}: Invalid parameter",
            -2 => $"{operation}: Memory allocation failed",
            -3 => $"{operation}: Random number generation failed",
            -4 => $"{operation}: Key generation failed",
            -5 => $"{operation}: Message value exceeds modulus",
            -6 => $"{operation}: Not initialized - call Initialize() first",
            -7 => $"{operation}: Invalid security level",
            _ => $"{operation}: Unknown error (code: {errorCode})"
        };
        return new KazKemException(message, errorCode);
    }
}

/// <summary>
/// Exception thrown when KAZ-KEM is not initialized.
/// </summary>
public class KazKemNotInitializedException : KazKemException
{
    public KazKemNotInitializedException()
        : base("KAZ-KEM is not initialized. Call KazKemContext.Initialize() first.", -6)
    {
    }
}

/// <summary>
/// Exception thrown when an invalid security level is specified.
/// </summary>
public class InvalidSecurityLevelException : KazKemException
{
    public SecurityLevel? AttemptedLevel { get; }

    public InvalidSecurityLevelException(int level)
        : base($"Invalid security level: {level}. Valid levels are 128, 192, or 256.", -7)
    {
        AttemptedLevel = null;
    }

    public InvalidSecurityLevelException(SecurityLevel level)
        : base($"Failed to initialize security level: {(int)level}", -7)
    {
        AttemptedLevel = level;
    }
}
